File: 0001-two-sum/0001-two-sum.js
/**
 * @param {number[]} nums
 * @param {number} target
 * @return {number[]}
 */

// approach 1) brute force
// look through each index in the array
// subtract target by the selected index number and store within variable
// check the rest of the numbers within the array
// if the result matches another number within the array, return the index of the selected number and the index of the matching number

// var twoSum = function(nums, target) {
//   for (let i = 0; i < nums.length; i++) {
//     const number = target - nums[i];
//     for (let x = i + 1; x < nums.length; x++) {
//       if (number === nums[x]) {
//         return [i, x];
//       }
//     }
//   }
// };


// brute force
// Time: O(n^2)
// Space: O(1)

// The time complexity of this solution is O(n^2), where n is the number of elements in the nums array. This is because there are nested loops where each iteration of the outer loop runs n times, and each iteration of the inner loop runs (n - i - 1) times on average, where i is the current index of the outer loop.

// The space complexity of this solution is O(1) because it only uses a constant amount of extra space regardless of the size of the input array.

// var twoSum = function(nums, target) {
//     for (let i = 0; i < nums.length; i++) {
//         const number = target - nums[i];
//         for (let x = i + 1; x < nums.length; x++) {
//             if (number === nums[x]) {
//                 return [i, x];
//             }
//         }
//     }
// }


// approach 2) hashmap
// Time: O(n)
// Memory: O(n)
// create an object literal for data storage
// look through each index within the array
// subtract target by the selected index number and store within variable
// if the stored value exists within the object, return the value and the index of the matching number
// if the stored value does not exist, use bracket notation to store the value of the element in the array as a key with its index as its value

// var twoSum = function(nums, target) {
//     let map = {};
//     for (let i = 0; i < nums.length; i++) {
//         const number = target - nums[i];
//         if (number in map) {
//             return [map[number], i]
//         }
//         map[nums[i]] = i;
//     }
// };


// var twoSum = function(nums, target) {
//     let map = {};
//     for (let i = 0; i < nums.length; i++) {
//         const number = target - nums[i];
//         if (number in map) {
//             return [map[number], i]
//         }
//         map[nums[i]] = i;
//     }
// }

var twoSum = function(nums, target) {
    let map = {};
    for (let i = 0; i < nums.length; i++) {
        const number = target - nums[i];
        if (number in map) {
            return [map[number], i]
        }
        map[nums[i]] = i;
    }
}

// { 
// 2:0
// }
// [0, 1]

File: 0001-two-sum/0001-two-sum.java

// Brute Force
// Time Complexity O(n^2)
// --> two nested loops, compare every pair
// Space Complexity O(1)
// --> no extra storage (just indices)

// PSEUDOCODE
// Loop through each element with index i
// Inside that, loop through every leter element with index j
// Check if nums[i] + nums[j] == target
// if yes, return [i, j]
// if no pairs match, return an empty array

// class Solution {
//     public int[] twoSum(int[] nums, int target) {
//         for (int i = 0; i < nums.length; i++) {
//             for (int j = i + 1; j < nums.length; j++) {
//                 if (nums[i] + nums[j] == target) {
//                     return new int[]{i, j};
//                 }
//             }
//         }
//         return new int[]{};
//     }
// }

// HashMap One-Pass
// Time Complexity: O(n)
// -> each number is only processed once
// Space Complexity: O(n)
// -> HashMap lookups (containsKey, get, put) are O(1) on average

// Create an empty map to keep track of numebrs already seen
// key will be number itself
// value will be the index where we saw it

// loop through array using index i
// check if the complement is already in the map
// if it is, number has been seen before and return with current number
// return empty array if not found


class Solution {
    public int[] twoSum(int[] nums, int target) {
        Map<Integer, Integer> map = new HashMap<>(); // stores value -> index

        for (int i = 0; i < nums.length; i++) {
            int complement = target - nums[i]; // what number we need to reach the target

            if (map.containsKey(complement)) { // check if we already saw the complement 
                return new int[] { map.get(complement), i }; // found the pair
            }

            map.put(nums[i], i); // store current number with its index
        }

        return new int[] {}; // no pair found (edge case)
    }
}


// class Solution {
//     public int[] twoSum(int[] nums, int target) {
//         Map<Integer, Integer> map = new Hashmap<>();

//         for (int i = 0; i < nums.length; i++) {
//             int number = target - nums[i];

//             if (map.containsKey(number)) {
//                 return new int[] { map.get(complement), i };
//             }

//             map.put(nums[i], i);
//         }

//         return new int[] {};
//     }
// }